### PR TITLE
MV2: replace BugDB with TrackerDB

### PR DESCRIPTION
**Pull request by chrmod**
TODO:
* [x] missing icon for category `Extensions`
   Screenshot 2023-03-07 at 17 30 52
* [x] missing translation `Telemetry`
* [x] `Unknown` category handling
  Screenshot 2023-03-07 at 17 30 55
* handle TrackerDB updates (on extension upgrades) 
  * [x] update `selected_app_ids` 
  * [x] how to deal with new trackers (without `ghostery_id`)
* TrackerDB detection and blocking
  * [x] decide if domain matching is needed
  * [x] consider disabling of blocking (as is provided by Ad-blocking already)

### CODE REVIEW

File: extension-manifest-v2/src/classes/BugDb.js [hunk 1]
```diff
@@ -10,246 +10,85 @@
...
-			// Bias in favor of blocking if the category will be
-			// blocked by default in a new installtion. This is
-			// mostly relevant if new categories were introduced.
-			majority[category] = 0.5;
-		});
-		Object.entries(old_bugs.apps).forEach(([app_id, { cat: category }]) => {
-			const vote = selected_app_ids[app_id] ? 1 : -1;
-			majority[category] = (majority[category] || 0) + vote;
-		});
+		this.db.bugs = engine.metadata.getPatterns().reduce((all, pattern) => ({
```
Comment (philipp-classen) [inline extension-manifest-v2/src/classes/BugDb.js:82]: The number of patterns isn't small, right? Be careful with the reduce-patterns. I think, it results in O(n²) complexity.

Comment (chrmod) [reply extension-manifest-v2/src/classes/BugDb.js]: fixed

**Review by smalluban — DISMISSED** [review]
I did not notice obvious bugs. However, are we fine with hardcoding a trackerdb in MV2, where before the bugs where fetched dynamically?

**Comment by philipp-classen** [comment]
> I did not notice obvious bugs. However, are we fine with hardcoding a trackerdb in MV2, where before the bugs where fetched dynamically?

Valid point (though I don't know what the implications of not updating will be).

As I understand, trackerdb can be seen as a link between community patterns and curated meta data. Meta data doesn't change frequently in my experience; it might be OK to update only on extension releases (even though from a maintainer perspective, it is preferable to be able to push fixes without releases).

In contrast to meta data, community list patterns (and trackers) change frequently. That means, it will naturally become more out-of-sync over time. Whether that will create problems or not, is hard to tell from my side. In any case, it would be desirable if the extension fails gracefully if there are inconsistencies. Maybe that mitigates the problem.

File: extension-manifest-v2/manifest.json [hunk 2]
```diff
@@ -3,8 +3,8 @@
 	"author": "Ghostery",
 	"name": "__MSG_name__",
 	"short_name": "Ghostery",
-	"version": "8.9.14",
-	"version_name": "8.9.14",
+	"version": "8.10.0",
```
Comment (smalluban) [inline extension-manifest-v2/manifest.json:6]: I would not bump the version in feature PR, it should be done in a separate PR with its own commit message (and then you have a clean commit, which you can tag).

File: package.json [hunk 3]
```diff
@@ -5,7 +5,7 @@
   "type": "module",
   "scripts": {
     "glossary-converter": "node tools/glossary-converter.js",
-    "postinstall-mv2": "cd extension-manifest-v2 && npm run licenses",
+    "postinstall-mv2": "cd extension-manifest-v2 && npm run licenses && npm run download-trackerdb",
```
Comment (smalluban) [inline package.json:8]: Is there a reason why packages postinstall scripts are not in the packages, but here? It was before like `npm postinstall -ws ...`.

File: ui/src/utils/categories.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ const colors = {
   essential: '#FC9734',
   hosting: '#8459A5',
   customer_interaction: '#EF671E',
-  unknown: '#E8E8E8',
+  unidentified: '#E8E8E8',
```
Comment (smalluban) [inline ui/src/utils/categories.js:19]: if we add a new category name, we must update locales in MV3 too.

**Review by smalluban — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (chrmod) [reply package.json:8]: Running `postinstall` from packages is not a supported usecase for npm workspaces. it started to fail for me with latest npm upgrade (when we've moved to node 19) - 

This approach is more predictable

File: extension-manifest-v2/_locales/de/messages.json [hunk 5]
```diff
@@ -537,17 +537,14 @@
   },
   "category_telemetry_desc": {
     "description": "User-facing description for the Telemetry tag category.",
-    "message": "Telemetrie"
+    "message": "category_advertising_desc"
```
Comment (philipp-classen) [inline extension-manifest-v2/_locales/de/messages.json:540]: Is this change intentional?

File: extension-manifest-v2/src/background.js [hunk 6]
```diff
@@ -100,24 +100,6 @@ function updateDBs() {
...
-				log('CHECK LIBRARY VERSION CALLED', result);
-				if (result.success) {
-					const nowTime = Date.now();
-					conf.bugs_last_checked = nowTime;
-					if (result.updated) {
-						log('BUGS LAST UPDATED UPDATED', new Date());
-						conf.bugs_last_updated = nowTime;
-					}
-				}
-				resolve({
```
Comment (philipp-classen) [inline extension-manifest-v2/src/background.js:113]: Won't removing that "resolve" leave the Promise that "updateDBs" returns dangling?

File: extension-manifest-v2/src/classes/BugDb.js [hunk 7]
```diff
@@ -10,246 +10,128 @@
...
-			categories[category].trackers.push({
-				id: appId,
-				name: db.apps[appId].name,
-				description: '',
-				blocked,
-				shouldShow: true,
-				catId: category,
-				trackerID: db.apps[appId].trackerID,
-			});
+	async init(just_upgraded) {
```
Comment (philipp-classen) [inline extension-manifest-v2/src/classes/BugDb.js:46]: Is it possible that this function gets called multiple times in parallel? If so, the engine might get loaded multiple times. (Which would not be a correctness problem, but would be wasteful.)

File: package.json (hunk 3 above)
Comment (smalluban) [reply package.json:8]: Ok, understand, but it ties the root with package internals... I get the reason but still don't like the solution.

**Review by smalluban — APPROVED** [review]
My requested changes added.

File: extension-manifest-v2/src/classes/BugDb.js (hunk 7 above)
Comment (chrmod) [reply extension-manifest-v2/src/classes/BugDb.js:46]: the only place it is being called is during background startup

File: extension-manifest-v2/_locales/de/messages.json (hunk 5 above)
Comment (chrmod) [reply extension-manifest-v2/_locales/de/messages.json:540]: nope

File: extension-manifest-v2/src/background.js (hunk 6 above)
Comment (chrmod) [reply extension-manifest-v2/src/background.js:113]: good catch

File: extension-manifest-v2/src/classes/BugDb.js [hunk 8]
```diff
@@ -10,246 +10,135 @@
...
-							selected_app_ids[app_id] = 1;
-						});
-
-						// (triggers a persist)
-						conf.selected_app_ids = selected_app_ids;
-					}
-					db.JUST_UPDATED_WITH_NEW_TRACKERS = true;
+		/* block new trackers if in mostly blocked category */
+		if (just_upgraded) {
+			Object.values(categoriesMeta).forEach((meta) => {
```
Comment (philipp-classen) [inline extension-manifest-v2/src/classes/BugDb.js:112]: Shouldn't we preserve the existing logic (_scanForNewTrackersToBlock) to take into account which categories are enabled/disabled by default?

A use case is when a new category is added, or if the extension for some reason started with an empty set of trackers in one category. As I understand the new code, the new trackers would be disabled now.

**Review by philipp-classen — APPROVED** [review]
